orderer/consensus/etcdraft/chain.go

```diff
@@ -921,10 +921,72 @@ func (c *Chain) catchUp(snap *raftpb.Snapshot) error {
 		next++
 	}
 
+	if memberChanged { // There is a membership change, need to re-calculate local raft id and refresh communications
+		id, err := c.detectSelfID(c.opts.Consenters)
+		if err != nil { // means i'm not in the channel, halt it
+			c.logger.Debugf("i'm not in the channel, will Halt", c.channelID)
+			c.Halt()
+			return nil
+		}
+		if c.raftID != id { // my id is changed
+			c.logger.Debugf("detectSelfID=%d, will update to it", c.channelID, id)
+
+			// Update to new id value
+			c.raftID = id
+			lg := c.opts.Logger.With("channel", c.channelID, "node", id)
+			c.logger = lg
+			c.opts.RaftID = id
+
+			c.Node.logger = lg
+			c.Node.config.ID = id
+			c.Node.config.Logger = lg
+			c.Node.metadata = c.opts.BlockMetadata
+			c.logger.Debugf("Updated to new ID=%d, will restart the raft.Node", c.channelID, id)
+
+			c.Node.Node.Stop()
+			c.Node.Node = raft.RestartNode(c.Node.config)
+		} else {
+			c.logger.Debugf("ID (%d) no change after catchup", c.channelID, id)
+		}
+
+		// refresh communications to other OSNs
+		if err := c.configureComm(); err != nil {
+			c.logger.Panicf("Failed to configure communication: %s", err)
+		}
+	}
+
 	c.logger.Infof("Finished syncing with cluster up to and including block [%d]", b.Header.Number)
 	return nil
 }
 
+// detectSelfID will try to find the current osn's raft id from the given consenters.
+func (c *Chain) detectSelfID(consenters map[uint64]*etcdraft.Consenter) (uint64, error) {
```

Comment:
Why are you copying the [method](https://github.com/hyperledger/fabric/blob/master/orderer/consensus/etcdraft/consenter.go#L106-L128) from the consenter? You can instead refactor the code to a non-instance method and pass the logger as a parameter.

Reply:
Well, that's not to break the consenter part. 
But It's OK, and I can change the code to move this dtectSelfID() method into some utils category. How about that?

Reply:
Besides, the current Chain struct cannot restart with new consensus config, and cannot re-apply its config once created. That's also part of reasons we need to carefully update these config items at runtime. The other reason is to improve performance, because a recreation of chain always take more time than updating its parameters.

Reply:
>  I can change the code to move this dtectSelfID() method into some utils category. How about that?

Sounds good.